sw/device/lib/base/memory_perftest.c

```diff
@@ -0,0 +1,319 @@
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#define PERFTEST_MAX_NUM_REPS 20
+#define PERFTEST_BUF_LEN 1000
+
+typedef struct test_context {
+  uint8_t buf1[PERFTEST_BUF_LEN];
+  uint8_t buf2[PERFTEST_BUF_LEN];
+  size_t len;
+} test_context_t;
+
+typedef struct perf_test {
+  // A human-readable name for this particular test, e.g. "memcpy".
+  const char *label;
+
+  // Setup functions for each context buffer. These function pointers may be
+  // NULL. The runtime of these functions will not be measured.
+  void (*setup_buf1)(uint8_t *, size_t);
+  void (*setup_buf2)(uint8_t *, size_t);
+
+  // Points to a function that exercises the function under test. This is the
+  // behavior that should be timed. It must not be NULL.
+  void (*func)(test_context_t *);
+
+  // The number of times to repeat this particular perf test. This is a
+  // per-test option because our 32-bit variance computation is sensitive to
+  // numerical overflow; tests that have a high variance should be run fewer
+  // times to avoid this.
+  size_t num_reps;
+} perf_test_t;
+
+static inline uint64_t read_cycle_counter();
+
+static void test_context_init(test_context_t *context) {
+  memset(context->buf1, 0, sizeof(context->buf1));
+  memset(context->buf2, 0, sizeof(context->buf2));
+  static_assert(sizeof(context->buf1) == PERFTEST_BUF_LEN, "");
+  static_assert(sizeof(context->buf2) == PERFTEST_BUF_LEN, "");
+  context->len = PERFTEST_BUF_LEN;
+}
+
+// Run the given `perf_test_t` and return the number of cycles it took.
+static inline uint64_t perf_test_run(test_context_t *context,
+                                     const perf_test_t *test) {
+  if (test->setup_buf1 != NULL) {
+    test->setup_buf1(context->buf1, context->len);
+  }
+  if (test->setup_buf2 != NULL) {
+    test->setup_buf2(context->buf2, context->len);
+  }
+  CHECK(test->func != NULL);
+
+  uint64_t start_cycles = read_cycle_counter();
+  test->func(context);
+  uint64_t end_cycles = read_cycle_counter();
+
+  // Check that executing the test function will take more than zero cycles.
+  // Practically, we should never see the 64-bit cycle counter overflow. Even if
+  // it took 12 hours to run all the perf tests, the clock would have to run at
+  // 427 THz to overflow the counter (2**64 cycles / (12 * 60 * 60) seconds).
+  CHECK(start_cycles < end_cycles);
+  return end_cycles - start_cycles;
+}
+
+static inline uint64_t read_cycle_counter() {
+#ifdef OT_PLATFORM_RV32
+  // We cannot read the entire 64-bit mcycle counter at once on RISCV32.
+  uint32_t mcycleh, mcycleh2, mcycle_lower;
+  do {
+    CSR_READ(CSR_REG_MCYCLEH, &mcycleh);
+    CSR_READ(CSR_REG_MCYCLE, &mcycle_lower);
+    CSR_READ(CSR_REG_MCYCLEH, &mcycleh2);
+  } while (mcycleh != mcycleh2);
+
+  return ((uint64_t)mcycleh << 32) | mcycle_lower;
+#else
+  static_assert(false, "Not implemented");
+#endif
+}
+
+static inline bool compute_statistics(const uint64_t *durations,
```

Comment:
Please see my comment on `num_reps`. If you agree we can park this in a gist until we need it.

Reply:
Resolving this — no need to keep track of the code for computing variance, it will live on in my memory :P